File: GoogleAnalytics.Win8/TransactionBuilder.cs
using Windows.ApplicationModel.Store;

namespace GoogleAnalytics
{
    public static class TransactionBuilder
    {
        static TransactionBuilder()
        {
            StoreName = "Windows 8 App Store";
        }

        /// <summary>
        /// Gets or sets the default strore name to be used when logging transactions
        /// </summary>
        public static string StoreName { get; set; }

        /// <summary>
        /// Constructs a transaction object with a transaction item for a product purchase
        /// </summary>
        /// <param name="listingInformation">The product listing information for the app</param>
        /// <param name="receipt">The receipt from the purchase</param>
        /// <returns>A transaction object all ready to get passed to Tracker.sendTransaction</returns>
        public static Transaction GetProductPurchaseTransaction(ListingInformation listingInformation, string receipt)
        {
            var productReceipt = ProductReceipt.Load(receipt);
            var transactionId = productReceipt.Id;
            var productId = productReceipt.ProductId;
            var productType = productReceipt.ProductType;

            var product = listingInformation.ProductListings[productId];
            var regionInfo = new System.Globalization.RegionInfo(listingInformation.CurrentMarket);
            var currencyCode = regionInfo.ISOCurrencySymbol;
            var currencyFormatter = new Windows.Globalization.NumberFormatting.CurrencyFormatter(currencyCode);
            var cost = currencyFormatter.ParseDouble(product.FormattedPrice);
            var costInMicrons = (long)(cost.GetValueOrDefault(0) * 1000000);

            var transaction = new Transaction(transactionId, costInMicrons);
            transaction.Affiliation = StoreName;
            transaction.CurrencyCode = currencyCode;
            var transactionItem = new TransactionItem(productId, product.Name, costInMicrons, 1);
            transactionItem.Category = productType;
            transaction.Items.Add(transactionItem);
            return transaction;
        }

        /// <summary>
        /// Constructs a transaction object with a transaction item for an app purchase
        /// </summary>
        /// <param name="listingInformation">The product listing information for the app</param>
        /// <param name="receipt">The receipt from the purchase</param>
        /// <returns>A transaction object all ready to get passed to Tracker.sendTransaction</returns>
        public static Transaction GetAppPurchaseTransaction(ListingInformation listingInformation, string receipt)
        {
            var appReceipt = AppReceipt.Load(receipt);
            var transactionId = appReceipt.Id;
            var appId = appReceipt.AppId;
            var licenseType = appReceipt.LicenseType;

            var regionInfo = new System.Globalization.RegionInfo(listingInformation.CurrentMarket);
            var currencyCode = regionInfo.ISOCurrencySymbol;
            var currencyFormatter = new Windows.Globalization.NumberFormatting.CurrencyFormatter(currencyCode);
            var cost = currencyFormatter.ParseDouble(listingInformation.FormattedPrice);
            var costInMicrons = (long)(cost.GetValueOrDefault(0) * 1000000);

            var transaction = new Transaction(transactionId, costInMicrons);
            transaction.Affiliation = StoreName;
            transaction.CurrencyCode = currencyCode;
            var transactionItem = new TransactionItem(appId, listingInformation.Name, costInMicrons, 1);
            transactionItem.Category = licenseType;
            transaction.Items.Add(transactionItem);
            return transaction;
        }
    }
}

File: GoogleAnalytics.Native/Common/TransactionBuilder.cpp
//
// TransactionBuilder.cpp
// Implementation of the TransactionBuilder class.
//

#include "pch.h"
#include "TransactionBuilder.h"
#include "ProductReceipt.h"
#include "AppReceipt.h"

using namespace GoogleAnalytics;
using namespace Platform;
using namespace Windows::ApplicationModel::Store;

String^ TransactionBuilder::storeName = "Windows 8 App Store";

Transaction^ TransactionBuilder::GetProductPurchaseTransaction(ListingInformation^ listingInformation, String^ receipt)
{
	auto productReceipt = ProductReceipt::Load(receipt);
	auto transactionId = productReceipt->Id;
	auto productId = productReceipt->ProductId;
	auto productType = productReceipt->ProductType;

	auto product = listingInformation->ProductListings->Lookup(productId);

	auto currencyCode = Windows::System::UserProfile::GlobalizationPreferences::Currencies->GetAt(0);

	auto currencyFormatter = ref new Windows::Globalization::NumberFormatting::CurrencyFormatter(currencyCode);
	auto cost = currencyFormatter->ParseDouble(product->FormattedPrice);
	auto costInMicrons = cost != nullptr ? (long long)(cost->Value * 1000000) : 0L;

	auto transaction = ref new Transaction(transactionId, costInMicrons);
	transaction->Affiliation = StoreName;
	transaction->CurrencyCode = currencyCode;
	auto transactionItem = ref new TransactionItem(productId, product->Name, costInMicrons, 1);
	transactionItem->Category = productType;
	transaction->Items->Append(transactionItem);
	return transaction;
}

Transaction^ TransactionBuilder::GetAppPurchaseTransaction(ListingInformation^ listingInformation, String^ receipt)
{
	auto appReceipt = AppReceipt::Load(receipt);
	auto transactionId = appReceipt->Id;
	auto appId = appReceipt->AppId;
	auto licenseType = appReceipt->LicenseType;
	
	auto currencyCode = Windows::System::UserProfile::GlobalizationPreferences::Currencies->GetAt(0);

	auto currencyFormatter = ref new Windows::Globalization::NumberFormatting::CurrencyFormatter(currencyCode);
	auto cost = currencyFormatter->ParseDouble(listingInformation->FormattedPrice);
	auto costInMicrons = cost != nullptr ? (long long)(cost->Value * 1000000) : 0L;

	auto transaction = ref new Transaction(transactionId, costInMicrons);
	transaction->Affiliation = StoreName;
	transaction->CurrencyCode = currencyCode;
	auto transactionItem = ref new TransactionItem(appId, listingInformation->Name, costInMicrons, 1);
	transactionItem->Category = licenseType;
	transaction->Items->Append(transactionItem);
	return transaction;
}


